src/internal/kopia/cleanup_backups.go

```diff
@@ -0,0 +1,157 @@
+	// probably happens O(minutes) or O(hours) instead of O(days). Of course, that
+	// assumes a non-adversarial setup where things such as machine hiberation,
+	// process freezing (i.e. paused at the OS level), etc. don't occur.
+
+	// Sort all the snapshots as either details snapshots or item data snapshots.
+	for _, snap := range snaps {
+		k, _ := makeTagKV(TagBackupCategory)
+		if _, ok := snap.Labels[k]; ok {
+			dataSnaps[snap.ID] = struct{}{}
+			continue
+		}
+
+		deets[snap.ID] = struct{}{}
+	}
+
+	// Get all legacy backup details models. The initial version of backup delete
+	// didn't seem to delete them so they may also be orphaned if the repo is old
+	// enough.
+	deetsModels, err := bs.GetIDsForType(ctx, model.BackupDetailsSchema, nil)
+	if err != nil {
+		return clues.Wrap(err, "getting legacy backup details")
+	}
+
+	for _, d := range deetsModels {
+		deets[d.ModelStoreID] = struct{}{}
+	}
+
+	// Get all backup models.
+	bups, err := bs.GetIDsForType(ctx, model.BackupSchema, nil)
+	if err != nil {
+		return clues.Wrap(err, "getting all backup models")
+	}
+
+	toDelete := maps.Clone(deets)
+	maps.Copy(toDelete, dataSnaps)
+
+	for _, bup := range bups {
+		toDelete[manifest.ID(bup.ModelStoreID)] = struct{}{}
+
+		bm := backup.Backup{}
+
+		if err := bs.GetWithModelStoreID(
+			ctx,
+			model.BackupSchema,
+			bup.ModelStoreID,
+			&bm,
+		); err != nil {
+			if !errors.Is(err, data.ErrNotFound) {
+				return clues.Wrap(err, "getting backup model").
+					With("search_backup_id", bup.ID)
+			}
+
+			// TODO(ashmrtn): This actually needs revised, see above TODO. Leaving it
+			// here for the moment to get the basic logic in.
+			//
+			// Safe to continue if the model wasn't found because that means that the
+			// possible item data and details for the backup are now orphaned. They'll
+			// be deleted since we won't remove them from the delete set.
+			//
+			// This isn't expected to really pop up, but it's possible if this
+			// function is run concurrently with either a backup delete or another
+			// instance of this function.
+			logger.Ctx(ctx).Debugw(
+				"backup model not found",
+				"search_backup_id", bup.ModelStoreID)
+
+			continue
+		}
+
+		ssid := bm.StreamStoreID
+		if len(ssid) == 0 {
+			ssid = bm.DetailsID
+		}
+
+		_, dataOK := dataSnaps[manifest.ID(bm.SnapshotID)]
+		_, deetsOK := deets[manifest.ID(ssid)]
+
+		if deetsOK && dataOK {
+			delete(toDelete, bup.ModelStoreID)
+			delete(toDelete, manifest.ID(bm.SnapshotID))
+			delete(toDelete, manifest.ID(ssid))
+
+			continue
+		}
+	}
+
+	// Use single atomic batch delete operation to cleanup to keep from making a
+	// bunch of manifest content blobs.
+	if err := bs.DeleteWithModelStoreIDs(ctx, maps.Keys(toDelete)...); err != nil {
+		return clues.Wrap(err, "deleting orphaned data")
+	}
+
+	// TODO(ashmrtn): Do some pruning of assist backup models so we don't keep
+	// them around forever.
+
+	return nil
```

Comment:
Maybe you plan to add this in a later PR, but should we log and/or return stats on deleted snapshots?

Reply:
ah yea, I actually missed that to start with and added it in an upcoming PR